Blocks: Insert hooked blocks

ockham opened the pull request:
## Synopsis

This feature is meant to provide an extensibility mechanism for Block Themes, in analogy to WordPress' [Hooks](https://developer.wordpress.org/plugins/hooks/) concept that has allowed extending Classic Themes through filters and actions.

Specifically, Block Hooks allow a third-party block to specify a position relative to a given block into which it will then be automatically inserted (e.g. a "Like" button block can ask to be inserted after the Post Content block, or an eCommerce shopping cart block can ask to be inserted after the Navigation block).

The two core tenets for block hooks are:

1. Insertion into the frontend should happen right after a plugin containing a hooked block is activated (i.e. the user isn't required to insert the block manually in the editor first); similarly, disabling the plugin should remove the hooked block from the frontend.
2. The user has the ultimate power to customize that automatic insertion: The hooked block is also visible in the editor, and the user's decision to persist, dismiss, customize, or move it will be respected (and reflected on the frontend).

To account for both tenets, we've made the **tradeoff** that automatic block insertion only works for unmodified templates (and template parts, respectively). The reason for this is that the simplest way of storing the information whether a block has been persisted to (or dismissed from) a given template (or part) is right in the template markup. This was first suggested [here](https://github.com/WordPress/gutenberg/issues/39439#issuecomment-1150278043).

To accommodate for that tradeoff, we've [added UI controls (toggles)](https://github.com/WordPress/gutenberg/pull/52969) to increase visibility of hooked blocks, and to allow their later insertion into templates (or parts) that already have been modified by the user.

## Implementation

Since we wanted hooked blocks to appear both in the frontend and in the editor (see tenet number 2), we had to make sure they were inserted into both the frontend markup and the REST API (templates and patterns endpoints) equally. As a consequence, this means that automatic insertion couldn't only be implemented at block _render_ stage, as for the editor, we needed to modify the _serialized_ (but _unrendered_) markup.

However, thanks to the tradeoff we made (see above), we could at least limit ourselves to only inserting hooked blocks into unmodified templates (and parts), i.e. those that were coming directly from a Block Theme's template (or parts) file, rather than the database, and patterns.

It turns out that there's a rather natural stage for automatic insertion of hooked blocks to happen, which is during template file retrieval (and loading of patterns, respectively).

Trac ticket: https://core.trac.wordpress.org/ticket/59313

Supersedes #5158.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

ockham commented:
@gziolo I'm hoping that this PR will become the final piece to implement Block Hooks. It should be pretty much feature-complete if I didn't miss anything 🙂

ockham commented on src/wp-includes/blocks.php:
```diff
@@ -760,6 +760,86 @@ function get_hooked_blocks( $name ) {
+ *
+ * @param WP_Block_Template|array $context A block template, template part, or pattern that the blocks belong to.
+ * @return callable A function that returns the serialized markup for the given block.
+ */
```
FIXME: Need proper PHPDoc.

ockham replied:
https://github.com/WordPress/wordpress-develop/pull/5261/commits/f6ad79a30b31678bd7698891a8c0b1180c7e7b75

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -760,6 +760,86 @@ function get_hooked_blocks( $name ) {
+		$hooked_blocks = get_hooked_blocks( $block['blockName'] );
+		foreach ( $hooked_blocks as $hooked_block_type => $relative_position ) {
+			if ( 'before' === $relative_position ) {
+				$hooked_block_markup = sprintf( '<!-- wp:%s /-->', strip_core_block_namespace( $hooked_block_type ) );
```
This could use the existing function `get_comment_delimited_block_content`:

https://github.com/WordPress/wordpress-develop/blob/58bb4b79ca8542d933cae26eafce32415fe5f8b8/src/wp-includes/blocks.php#L908

It would look something like:

```PHP
$hooked_block_markup = get_comment_delimited_block_content( $hooked_block_type, array(), '' );
```

It's mostly useful to hide the details of how the block delimiters are generated. I don't expect them to change, but that glues everything together.

ockham replied:
826837e2a2

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -760,6 +760,86 @@ function get_hooked_blocks( $name ) {
+	};
+}
+
+function make_after_block_visitor( $block_template ) {
```
Should it also take `$context` as the param name? That would be useful mostly for consistency. I also see now that it's used with patterns.

ockham replied:
Oh yeah, good point. Oversight on my part.

ockham replied:
8adc7a7d21

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -760,6 +760,86 @@ function get_hooked_blocks( $name ) {
+				 * @param array                   $block               The anchor block.
+				 * @param WP_Block_Template|array $context             The block template, template part, or pattern that the anchor block belongs to.
+				 */
+				$markup .= apply_filters( 'insert_hooked_block', $hooked_block_markup, $hooked_block_type, $relative_position, $block, $context );
```
This is nice, so you essentially do:

```php
add_filter( 'insert_hooked_block', function( $markup, $block_type, $position ) {
    if ( 'tests/some-block' === $block_type && 'after' === $position ) {
        return '';
    }
    return $markup;
}, 10, 3 );
```

In case you want to prevent injecting a block matching your conditions. It can also cover the current block and the context: pattern or template. Powerful!

ockham replied:
Exactly, that's the idea!

One thing we might still want to add (in a follow-up) could be a similar filter to allow _adding_ a block conditionally 🤔 (So it doesn't need to be registered in a `block.json` first.)

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -760,6 +760,86 @@ function get_hooked_blocks( $name ) {
+
+		if ( $parent && ! $prev ) {
+			// Candidate for first-child insertion.
+			$hooked_blocks_for_parent = get_hooked_blocks( $parent['blockName'] );
```
`get_hooked_blocks` is going to be called too many times. I will work on some optimizations tomorrow.

ockham commented:
Weird, unit tests are currently failing:

```
Invalid argument supplied for foreach()

/var/www/src/wp-includes/blocks.php:754
```

That's this line: https://github.com/WordPress/wordpress-develop/blob/aa033cba5cd3424a0445c6cd95e34307986880a1/src/wp-includes/blocks.php#L754

I wonder if/why the `block_hooks` field isn't available at all (rather than defaulting to the empty array)?

ockham commented:
Workaround in bd1c57bb20. Let's see if that works.

ockham commented on src/wp-includes/blocks.php:
```diff
@@ -760,6 +763,86 @@ function get_hooked_blocks( $name ) {
+				 * @param array                   $block               The anchor block.
+				 * @param WP_Block_Template|array $context             The block template, template part, or pattern that the anchor block belongs to.
+				 */
+				$markup .= apply_filters( 'insert_hooked_block', $hooked_block_markup, $hooked_block_type, $relative_position, $block, $context );
```
Thinking to rename the filter to include the word "markup" in case we want to introduce a different filter that allows including a hooked block at an earlier stage (e.g. adding just the block type to the list of blocks returned from `get_hooked_blocks()`).

Maybe

```suggestion
				$markup .= apply_filters( 'insert_hooked_block_markup', $hooked_block_markup, $hooked_block_type, $relative_position, $block, $context );
```

or even

```suggestion
				$markup .= apply_filters( 'inject_hooked_block_markup', $hooked_block_markup, $hooked_block_type, $relative_position, $block, $context );
```

ockham replied:
29a42fb500

gziolo commented on src/wp-includes/blocks.php:
```diff
@@ -751,6 +751,9 @@ function get_hooked_blocks( $name ) {
 	$block_types = WP_Block_Type_Registry::get_instance()->get_all_registered();
 	$hooked_blocks = array();
 	foreach ( $block_types as $block_type ) {
+		if ( ! property_exists( $block_type, 'block_hooks' ) ) {
```
> I wonder if/why the block_hooks field isn't available at all (rather than defaulting to the empty array)?

It might be the tests for REST API that try to inject invalid values. `! empty( $block_type->block_hooks )` check would be faster and it would also cover the case when there is an empty array.

gziolo replied:
It might be an issue with the missing test clean up. I'm trying unregistering the block with invalid data: https://github.com/WordPress/wordpress-develop/pull/5261/commits/d78370204077268a39ed5ffc30f053abcf0dbb48.

ockham commented:
Seeing the same test failure I had [earlier](https://github.com/WordPress/wordpress-develop/pull/5261#issuecomment-1729263616):

```
Invalid argument supplied for foreach()

/var/www/src/wp-includes/blocks.php:757
```

Maybe I need to add an `is_array` check alongside `property_exists`.

ockham commented:
> Seeing the same test failure I had [earlier](https://github.com/WordPress/wordpress-develop/pull/5261#issuecomment-1729263616):
> 
> ```
> Invalid argument supplied for foreach()
> 
> /var/www/src/wp-includes/blocks.php:757
> ```
> 
> Maybe I need to add an `is_array` check alongside `property_exists`.

https://github.com/WordPress/wordpress-develop/pull/5261/commits/22c75d9e338b5ab021e5f481c0e389136b05eb70

gziolo reviewed (approved):
We still need to increase the code coverage with unit tests, but overall, it works as expected based on the smoke tests performed. I'm very confident that `traverse_and_serialize_blocks` is implemented correctly. Some follow-up work will require:

- Increased test coverage for  `make_before_block_visitor` and `make_after_block_visitor` (they are tested indirectly to some degree).
- Integration test for Block Hooks that spans a simplified theme based on Twenty Twenty-Three that contains template file with a template part and pattern sourced from the theme folders.
- Performance audit for `get_hooked_blocks` and potentially also to `traverse_and_serialize_blocks`.

ockham commented:
Thanks a lot for the test coverage @gziolo, I quite like your approach! 😄

ockham committed:
Blocks: Implement automatic block insertion into Block Hooks.

Block Hooks allow a third-party block to specify a position relative to a given block into which it will then be automatically inserted (e.g. a "Like" button block can ask to be inserted after the Post Content block, or an eCommerce shopping cart block can ask to be inserted after the Navigation block).

The underlying idea is to provide an extensibility mechanism for Block Themes, in analogy to WordPress' [https://developer.wordpress.org/plugins/hooks/ Hooks] concept that has allowed extending Classic Themes through filters and actions.

The two core tenets for Block Hooks are:

1. Insertion into the frontend should happen right after a plugin containing a hooked block is activated (i.e. the user isn't required to insert the block manually in the editor first); similarly, disabling the plugin should remove the hooked block from the frontend.
2. The user has the ultimate power to customize that automatic insertion: The hooked block is also visible in the editor, and the user's decision to persist, dismiss (i.e. remove), customize, or move it will be respected (and reflected on the frontend).

To account for both tenets, the **tradeoff** was made to limit automatic block insertion to unmodified templates (and template parts, respectively). The reason for this is that the simplest way of storing the information whether a block has been persisted to (or dismissed from) a given template (or part) is right in the template markup.

To accommodate for that tradeoff, [https://github.com/WordPress/gutenberg/pull/52969 UI controls (toggles)] are being added to increase visibility of hooked blocks, and to allow for their later insertion into templates (or parts) that already have been modified by the user.

For hooked blocks to appear both in the frontend and in the editor (see tenet number 2), they need to be inserted into both the frontend markup and the REST API (templates and patterns endpoints) equally. As a consequence, this means that automatic insertion couldn't (only) be implemented at block ''render'' stage, as for the editor, the ''serialized'' (but ''unrendered'') markup needs to be modified.

Furthermore, hooked blocks also have to be inserted into block patterns. Since practically no filters exist for the patterns registry, this has to be done in the registry's `get_registered` and `get_all_registered` methods.

Props gziolo.
Fixes #59313.

ockham commented:
Committed to Core in https://core.trac.wordpress.org/changeset/56649.
